File: tests/runners/go/tests/07_indexes_test.go
package polygen

import (
	"os"
	"path/filepath"
	"testing"
)

func TestIndexesAndForeignKeyValidation(t *testing.T) {
	container := NewSchemaContainer()
	user := &User{Id: 1, Username: "alice", Email: "alice@example.com", DisplayName: "Alice"}
	description := "Binary reference systems"
	category := &Category{Id: 10, Name: "news", Description: &description, Rank: 7, Kind: CategoryKindPublic}
	post := &Post{Id: 100, Title: "title", Content: "body", AuthorId: user.Id, CategoryId: category.Id}
	tag := &Tag{Id: 200, Name: "featured"}
	postTag := &PostTag{PostId: post.Id, TagId: tag.Id}

	container.Users.AddRow(user)
	container.Categorys.AddRow(category)
	container.Posts.AddRow(post)
	container.Tags.AddRow(tag)
	container.PostTags.AddRow(postTag)

	if container.Users.GetByUsername("alice") != user {
		t.Fatalf("unique username index did not return inserted row")
	}
	if got := container.Posts.GetByAuthorId(user.Id); len(got) != 1 || got[0] != post {
		t.Fatalf("group author index returned %#v", got)
	}
	if got := container.PostTags.GetByTagId(tag.Id); len(got) != 1 || got[0] != postTag {
		t.Fatalf("junction tag index returned %#v", got)
	}
	if got := container.GetPostAuthor(post); got != user {
		t.Fatalf("post author navigation returned %#v", got)
	}
	if got := container.GetPostCategory(post); got != category {
		t.Fatalf("post category navigation returned %#v", got)
	}
	if got := container.GetPostTagPost(postTag); got != post {
		t.Fatalf("post tag post navigation returned %#v", got)
	}
	if got := container.GetPostTagTag(postTag); got != tag {
		t.Fatalf("post tag tag navigation returned %#v", got)
	}
	if err := container.ValidateOrError(); err != nil {
		t.Fatalf("expected valid foreign keys: %v", err)
	}
}

func TestContainerSearchIndexes(t *testing.T) {
	container := NewSchemaContainer()
	description := "Binary reference systems"
	category := &Category{Id: 10, Name: "Technology", Description: &description, Rank: 7, Kind: CategoryKindPublic}
	post := &Post{Id: 100, Title: "Binary Reference Guide", Content: "body", AuthorId: 1, CategoryId: category.Id}
	other := &Post{Id: 101, Title: "General Indexes", Content: "body", AuthorId: 1, CategoryId: category.Id}

	container.Categorys.AddRow(category)
	container.Posts.AddRow(post)
	container.Posts.AddRow(other)

	if got := container.Categorys.SearchByName("technology"); len(got) != 1 || got[0] != category {
		t.Fatalf("exact string search returned %#v", got)
	}
	if got := container.Categorys.SearchByDescription("reference"); len(got) != 1 || got[0] != category {
		t.Fatalf("optional string token search returned %#v", got)
	}
	if got := container.Categorys.SearchByRank(7); len(got) != 1 || got[0] != category {
		t.Fatalf("numeric exact search returned %#v", got)
	}
	if got := container.Categorys.SearchByKind(CategoryKindPublic); len(got) != 1 || got[0] != category {
		t.Fatalf("enum exact search returned %#v", got)
	}
	if got := container.Posts.SearchByTitle("binary"); len(got) != 1 || got[0] != post {
		t.Fatalf("post title token search returned %#v", got)
	}
	if got := container.Posts.SearchByTitle("missing"); len(got) != 0 {
		t.Fatalf("missing token search returned %#v", got)
	}
	container.Clear()
	if got := container.Posts.SearchByTitle("binary"); len(got) != 0 {
		t.Fatalf("clear should remove search postings, got %#v", got)
	}
}

func TestBinaryRefSearchIndexes(t *testing.T) {
	description := "Binary reference systems"
	category := &Category{Id: 10, Name: "Technology", Description: &description, Rank: 7, Kind: CategoryKindPublic}
	post := &Post{Id: 100, Title: "Binary Reference Guide", Content: "body", AuthorId: 1, CategoryId: category.Id}
	other := &Post{Id: 101, Title: "General Indexes", Content: "body", AuthorId: 1, CategoryId: category.Id}

	data, err := SaveSchemaBinaryRefContext(&SchemaBinaryRefContainer{
		Users: []*User{
			{Id: 1, Username: "alice", Email: "alice@example.com", DisplayName: "Alice"},
		},
		Categorys: []*Category{category},
		Posts:     []*Post{post, other},
		Comments:  []*Comment{},
		Tags:      []*Tag{},
		PostTags:  []*PostTag{},
	})
	if err != nil {
		t.Fatalf("save binary ref: %v", err)
	}

	ctx, err := OpenSchemaBinaryRefContext(data)
	if err != nil {
		t.Fatalf("open binary ref: %v", err)
	}

	if got := ctx.Categorys.SearchByName("technology"); len(got) != 1 {
		t.Fatalf("binary ref exact string search returned %#v", got)
	} else {
		name, err := got[0].Name()
		if err != nil || name != "Technology" {
			t.Fatalf("binary ref name getter = %q, %v", name, err)
		}
	}
	if got := ctx.Categorys.SearchByDescription("reference"); len(got) != 1 {
		t.Fatalf("binary ref optional string token search returned %#v", got)
	}
	if got := ctx.Categorys.SearchByRank(7); len(got) != 1 {
		t.Fatalf("binary ref numeric exact search returned %#v", got)
	} else {
		rank, err := got[0].Rank()
		if err != nil || rank != 7 {
			t.Fatalf("binary ref rank getter = %d, %v", rank, err)
		}
	}
	if got := ctx.Categorys.SearchByKind(CategoryKindPublic); len(got) != 1 {
		t.Fatalf("binary ref enum exact search returned %#v", got)
	}
	if got := ctx.Posts.SearchByTitle("binary"); len(got) != 1 {
		t.Fatalf("binary ref post title token search returned %#v", got)
	} else {
		title, err := got[0].Title()
		if err != nil || title != "Binary Reference Guide" {
			t.Fatalf("binary ref title getter = %q, %v", title, err)
		}
	}
	if got := ctx.Posts.SearchByTitle("missing"); len(got) != 0 {
		t.Fatalf("binary ref missing token search returned %#v", got)
	}
}

func TestBinaryRefRejectsInvalidEnumWrite(t *testing.T) {
	description := "Invalid enum category"
	_, err := SaveSchemaBinaryRefContext(&SchemaBinaryRefContainer{
		Users: []*User{
			{Id: 1, Username: "alice", Email: "alice@example.com", DisplayName: "Alice"},
		},
		Categorys: []*Category{
			{Id: 10, Name: "InvalidKind", Description: &description, Rank: 1, Kind: CategoryKind(999)},
		},
		Posts:    []*Post{},
		Comments: []*Comment{},
		Tags:     []*Tag{},
		PostTags: []*PostTag{},
	})
	if err == nil {
		t.Fatalf("expected invalid BinaryRef enum write to fail")
	}
}

func TestGeneratedLoaders(t *testing.T) {
	dir := t.TempDir()
	write := func(name string, content string) {
		t.Helper()
		if err := os.WriteFile(filepath.Join(dir, name), []byte(content), 0644); err != nil {
			t.Fatalf("write %s: %v", name, err)
		}
	}

	write("users.csv", "id,username,email,display_name\n1,alice,alice@example.com,Alice\n")
	write("categories.csv", "id,name,description,rank,kind\n10,Technology,Binary reference systems,7,Public\n11,Internal,Private systems,3,CategoryKindInternal\n")
	write("posts.csv", "id,title,content,author_id,category_id\n100,Binary Reference Guide,body,1,10\n")
	write("comments.csv", "id,post_id,author_id,content,parent_id\n1,100,1,hello,\n")
	write("tags.csv", "id,name\n200,featured\n")
	write("post_tags.csv", "post_id,tag_id\n100,200\n")

	container := NewSchemaContainer()
	if err := container.LoadFromCsv(dir); err != nil {
		t.Fatalf("LoadFromCsv: %v", err)
	}
	if got := container.Users.Count(); got != 1 {
		t.Fatalf("loaded users = %d", got)
	}
	if got := container.Categorys.SearchByKind(CategoryKindInternal); len(got) != 1 || got[0].Name != "Internal" {
		t.Fatalf("loaded enum name search returned %#v", got)
	}
	if got := container.Posts.SearchByTitle("binary"); len(got) != 1 || got[0].Title != "Binary Reference Guide" {
		t.Fatalf("loaded post search returned %#v", got)
	}
	if err := container.ValidateOrError(); err != nil {
		t.Fatalf("loaded CSV data should validate: %v", err)
	}

	write("users.json", `[{"id":1,"username":"alice","email":"alice@example.com","display_name":"Alice"}]`)
	write("categories.json", `[{"id":12,"name":"JsonInternal","description":"Private systems","rank":4,"kind":"Internal"},{"id":13,"name":"JsonPublic","description":"Public systems","rank":5,"kind":1}]`)
	write("posts.json", `[{"id":101,"title":"JSON Binary Reference Guide","content":"body","author_id":1,"category_id":12}]`)
	write("comments.json", `[{"id":2,"post_id":101,"author_id":1,"content":"json hello","parent_id":null}]`)
	write("tags.json", `[{"id":201,"name":"json-tag"}]`)
	write("post_tags.json", `[{"post_id":101,"tag_id":201}]`)
	sourcesJsonContainer := NewSchemaContainer()
	if err := sourcesJsonContainer.LoadFromJson(dir); err != nil {
		t.Fatalf("LoadFromJson: %v", err)
	}
	if got := sourcesJsonContainer.Users.Count(); got != 1 {
		t.Fatalf("loaded JSON users = %d", got)
	}
	if got := sourcesJsonContainer.Categorys.SearchByKind(CategoryKindInternal); len(got) != 1 || got[0].Name != "JsonInternal" {
		t.Fatalf("sources JSON enum name search returned %#v", got)
	}
	if got := sourcesJsonContainer.Categorys.SearchByKind(CategoryKindPublic); len(got) != 1 || got[0].Name != "JsonPublic" {
		t.Fatalf("sources JSON numeric enum search returned %#v", got)
	}
	if got := sourcesJsonContainer.Posts.SearchByTitle("json binary"); len(got) != 1 || got[0].Title != "JSON Binary Reference Guide" {
		t.Fatalf("sources JSON post search returned %#v", got)
	}
	post := sourcesJsonContainer.Posts.GetById(101)
	if post == nil {
		t.Fatalf("sources JSON post id lookup failed")
	}
	if got := sourcesJsonContainer.GetPostAuthor(post); got == nil || got.Username != "alice" {
		t.Fatalf("sources JSON post author navigation returned %#v", got)
	}
	if got := sourcesJsonContainer.PostTags.GetByTagId(201); len(got) != 1 || got[0].PostId != 101 {
		t.Fatalf("sources JSON post tag index returned %#v", got)
	}
	if err := sourcesJsonContainer.ValidateOrError(); err != nil {
		t.Fatalf("loaded sources JSON data should validate: %v", err)
	}

	write("Category.json", `[{"id":12,"name":"JsonInternal","description":"Private systems","rank":4,"kind":"Internal"},{"id":13,"name":"JsonPublic","description":"Public systems","rank":5,"kind":1}]`)
	write("Tag.json", `[{"id":201,"name":"json-tag"}]`)
	jsonContainer := NewSchemaContainer()
	if err := jsonContainer.LoadFromJsonDirectory(dir); err != nil {
		t.Fatalf("LoadFromJsonDirectory: %v", err)
	}
	if got := jsonContainer.Categorys.SearchByKind(CategoryKindInternal); len(got) != 1 || got[0].Name != "JsonInternal" {
		t.Fatalf("loaded JSON enum name search returned %#v", got)
	}
	if got := jsonContainer.Categorys.SearchByKind(CategoryKindPublic); len(got) != 1 || got[0].Name != "JsonPublic" {
		t.Fatalf("loaded JSON numeric enum search returned %#v", got)
	}
	if got := jsonContainer.Tags.GetByName("json-tag"); got == nil || got.Id != 201 {
		t.Fatalf("loaded JSON tag lookup returned %#v", got)
	}

	write("bad_categories.csv", "id,name,description,rank,kind\n11,Bad,,not-a-number,1\n")
	if _, err := LoadCategorysFromCsv(filepath.Join(dir, "bad_categories.csv")); err == nil {
		t.Fatalf("expected invalid CSV numeric value to return an error")
	}

	write("bad_category_kind.csv", "id,name,description,rank,kind\n12,BadKind,,1,MissingKind\n")
	if _, err := LoadCategorysFromCsv(filepath.Join(dir, "bad_category_kind.csv")); err == nil {
		t.Fatalf("expected invalid CSV enum name to return an error")
	}

	write("bad_category_kind.json", `[{"id":14,"name":"BadJsonKind","description":"","rank":1,"kind":"MissingKind"}]`)
	if _, err := LoadCategorysFromJson(filepath.Join(dir, "bad_category_kind.json")); err == nil {
		t.Fatalf("expected invalid JSON enum name to return an error")
	}
}

func TestForeignKeyValidationRejectsMissingReference(t *testing.T) {
	container := NewSchemaContainer()
	container.Users.AddRow(&User{Id: 1, Username: "alice", Email: "alice@example.com", DisplayName: "Alice"})
	container.Posts.AddRow(&Post{Id: 100, Title: "title", Content: "body", AuthorId: 1, CategoryId: 999})

	result := container.ValidateAll()
	if result.IsValid() {
		t.Fatalf("expected missing category foreign key to be rejected")
	}
	if result.ErrorCount() != 1 {
		t.Fatalf("expected one validation error, got %d: %s", result.ErrorCount(), result.String())
	}
}

File: tests/runners/python/tests/07_indexes_test.py
import csv
import json
import struct
import tempfile
from pathlib import Path

from .schema import (
    TestIndexesCategory,
    TestIndexesCategoryKind,
    TestIndexesPost,
    TestIndexesPostTag,
    TestIndexesTag,
    TestIndexesUser,
    load_test_indexes_categorys_from_csv,
    load_test_indexes_categorys_from_json,
)
from .schema_binary_refs import BinaryRefDocument, SchemaBinaryRefContext, SchemaContainer as BinaryRefContainer
from .schema_container import SchemaContainer, ValidationException


def write_csv(path: Path, rows: list[dict[str, object]]) -> None:
    with path.open("w", encoding="utf-8", newline="") as handle:
        writer = csv.DictWriter(handle, fieldnames=list(rows[0].keys()))
        writer.writeheader()
        for row in rows:
            writer.writerow(row)


def read_i32(data: bytes | bytearray, offset: int) -> int:
    return struct.unpack_from("<i", data, offset)[0]


def patch_category_kind_payload(document: BinaryRefDocument, value: int) -> bytes:
    data = bytearray(document.to_bytes())
    row_offset = document.categorys.at(0)._row_offset
    kind_field_index = 4
    kind_relative_offset = read_i32(data, row_offset + 4 + kind_field_index * 4)
    if kind_relative_offset < 0:
        raise AssertionError("category kind field should be present")
    struct.pack_into("<i", data, row_offset + kind_relative_offset, value)
    return bytes(data)


def test_indexes_and_foreign_keys() -> None:
    container = SchemaContainer()
    user = TestIndexesUser(id=1, username="alice", email="alice@example.com", display_name="Alice")
    category = TestIndexesCategory(
        id=10,
        name="Technology",
        description="Binary reference systems",
        rank=7,
        kind=TestIndexesCategoryKind.Public,
    )
    post = TestIndexesPost(id=100, title="Binary Reference Guide", content="body", author_id=1, category_id=10)
    tag = TestIndexesTag(id=200, name="featured")
    post_tag = TestIndexesPostTag(post_id=100, tag_id=200)

    container.users.add_row(user)
    container.categorys.add_row(category)
    container.posts.add_row(post)
    container.tags.add_row(tag)
    container.post_tags.add_row(post_tag)

    assert container.users.get_by_username("alice") is user
    assert container.categorys.get_by_name("Technology") is category
    assert container.posts.find_by_author_id(1) == [post]
    assert container.post_tags.find_by_tag_id(200) == [post_tag]
    assert container.get_post_author(post) is user
    assert container.get_post_category(post) is category
    assert container.get_post_tag_post(post_tag) is post
    assert container.get_post_tag_tag(post_tag) is tag
    assert container.categorys.search_by_name(" technology ") == [category]
    assert container.categorys.search_by_description("reference") == [category]
    assert container.categorys.search_by_rank(7) == [category]
    assert container.categorys.search_by_kind(TestIndexesCategoryKind.Public) == [category]
    assert container.posts.search_by_title("binary") == [post]
    assert container.posts.search_by_title("missing") == []
    assert container.validate_all().is_valid()
    container.validate_or_raise()


def test_foreign_key_validation_rejects_missing_reference() -> None:
    container = SchemaContainer()
    container.users.add_row(TestIndexesUser(id=1, username="alice", email="alice@example.com", display_name="Alice"))
    container.posts.add_row(TestIndexesPost(id=100, title="bad", content="body", author_id=1, category_id=999))

    result = container.validate_all()
    assert not result.is_valid()
    assert result.error_count() == 1
    assert result.errors[0].field_name == "category_id"

    try:
        container.validate_or_raise()
    except ValidationException as exc:
        assert exc.result is result or exc.result.error_count() == 1
    else:
        raise AssertionError("validate_or_raise should reject missing foreign key")


def test_clear_resets_indexes() -> None:
    container = SchemaContainer()
    user = TestIndexesUser(id=1, username="alice", email="alice@example.com", display_name="Alice")
    container.users.add_row(user)
    assert container.users.get_by_id(1) is user
    container.clear()
    assert container.users.count() == 0
    assert container.users.get_by_id(1) is None


def test_enum_csv_json_loaders() -> None:
    with tempfile.TemporaryDirectory() as temp_dir:
        root = Path(temp_dir)
        json_path = root / "categories.json"
        csv_path = root / "categories.csv"

        json_path.write_text(
            json.dumps(
                [
                    {
                        "id": 1,
                        "name": "Public",
                        "description": "Visible category",
                        "rank": 5,
                        "kind": "Public",
                    }
                ]
            ),
            encoding="utf-8",
        )
        write_csv(
            csv_path,
            [
                {
                    "id": "2",
                    "name": "Internal",
                    "description": "Private category",
                    "rank": "8",
                    "kind": "2",
                }
            ],
        )

        from_json = load_test_indexes_categorys_from_json(json_path)
        from_csv = load_test_indexes_categorys_from_csv(csv_path)

    assert from_json == [
        TestIndexesCategory(
            id=1,
            name="Public",
            description="Visible category",
            rank=5,
            kind=TestIndexesCategoryKind.Public,
        )
    ]
    assert from_csv == [
        TestIndexesCategory(
            id=2,
            name="Internal",
            description="Private category",
            rank=8,
            kind=TestIndexesCategoryKind.Internal,
        )
    ]


def test_container_load_from_sources_csv() -> None:
    with tempfile.TemporaryDirectory() as temp_dir:
        root = Path(temp_dir)
        write_csv(
            root / "users.csv",
            [{"id": 1, "username": "alice", "email": "alice@example.com", "display_name": "Alice"}],
        )
        write_csv(
            root / "categories.csv",
            [{"id": 10, "name": "Technology", "description": "Binary reference systems", "rank": 7, "kind": "Public"}],
        )
        write_csv(
            root / "posts.csv",
            [{"id": 100, "title": "Binary Reference Guide", "content": "body", "author_id": 1, "category_id": 10}],
        )
        write_csv(
            root / "comments.csv",
            [{"id": 1000, "post_id": 100, "author_id": 1, "content": "nice", "parent_id": ""}],
        )
        write_csv(root / "tags.csv", [{"id": 200, "name": "featured"}])
        write_csv(root / "post_tags.csv", [{"post_id": 100, "tag_id": 200}])

        container = SchemaContainer()
        container.load_from_csv(root)

    post = container.posts.get_by_id(100)
    assert post is not None
    assert container.users.get_by_username("alice") is not None
    assert container.categorys.search_by_name(" technology ") == [container.categorys.get_by_id(10)]
    assert container.posts.search_by_title("binary") == [post]
    assert container.get_post_author(post).username == "alice"
    assert container.validate_all().is_valid()


def test_container_load_from_sources_json() -> None:
    with tempfile.TemporaryDirectory() as temp_dir:
        root = Path(temp_dir)
        (root / "users.json").write_text(
            json.dumps(
                [{"id": 1, "username": "alice", "email": "alice@example.com", "display_name": "Alice"}]
            ),
            encoding="utf-8",
        )
        (root / "categories.json").write_text(
            json.dumps(
                [
                    {
                        "id": 10,
                        "name": "Technology",
                        "description": "Binary reference systems",
                        "rank": 7,
                        "kind": "Public",
                    }
                ]
            ),
            encoding="utf-8",
        )
        (root / "posts.json").write_text(
            json.dumps(
                [{"id": 100, "title": "Binary Reference Guide", "content": "body", "author_id": 1, "category_id": 10}]
            ),
            encoding="utf-8",
        )
        (root / "comments.json").write_text(
            json.dumps([{"id": 1000, "post_id": 100, "author_id": 1, "content": "nice", "parent_id": None}]),
            encoding="utf-8",
        )
        (root / "tags.json").write_text(json.dumps([{"id": 200, "name": "featured"}]), encoding="utf-8")
        (root / "post_tags.json").write_text(json.dumps([{"post_id": 100, "tag_id": 200}]), encoding="utf-8")

        container = SchemaContainer()
        container.load_from_json(root)

    post = container.posts.get_by_id(100)
    assert post is not None
    assert container.users.get_by_username("alice") is not None
    assert container.categorys.search_by_name(" technology ") == [container.categorys.get_by_id(10)]
    assert container.posts.search_by_title("binary") == [post]
    assert container.get_post_author(post).username == "alice"
    assert container.validate_all().is_valid()


def test_binary_roundtrip_and_invalid_enum_discriminant() -> None:
    category = TestIndexesCategory(
        id=10,
        name="Technology",
        description="Binary reference systems",
        rank=7,
        kind=TestIndexesCategoryKind.Public,
    )
    loaded_category = TestIndexesCategory.from_binary(category.to_binary())
    assert loaded_category == category

    post = TestIndexesPost(id=100, title="Binary Reference Guide", content="body", author_id=1, category_id=10)
    loaded_post = TestIndexesPost.from_binary(post.to_binary())
    assert loaded_post == post

    payload = bytearray(category.to_binary())
    payload[-4:] = struct.pack("<i", 99)
    try:
        TestIndexesCategory.from_binary(payload)
    except ValueError:
        pass
    else:
        raise AssertionError("invalid enum discriminant should be rejected")

    invalid_category = TestIndexesCategory(
        id=11,
        name="Invalid",
        description=None,
        rank=1,
        kind=99,
    )
    try:
        invalid_category.to_binary()
    except ValueError:
        pass
    else:
        raise AssertionError("invalid enum discriminant should be rejected on write")


def test_binary_ref_document_roundtrip_and_lazy_indexes() -> None:
    container = SchemaContainer()
    user = TestIndexesUser(id=1, username="alice", email="alice@example.com", display_name="Alice")
    category = TestIndexesCategory(
        id=10,
        name="Technology",
        description="Binary reference systems",
        rank=7,
        kind=TestIndexesCategoryKind.Public,
    )
    post = TestIndexesPost(id=100, title="Binary Reference Guide", content="body", author_id=1, category_id=10)
    tag = TestIndexesTag(id=200, name="featured")
    post_tag = TestIndexesPostTag(post_id=100, tag_id=200)

    container.users.add_row(user)
    container.categorys.add_row(category)
    container.posts.add_row(post)
    container.tags.add_row(tag)
    container.post_tags.add_row(post_tag)

    document = BinaryRefDocument.from_container(container)
    assert document.users.count() == 1
    assert document.users.get_by_id(1).get() == user
    assert document.users.get_by_username("alice").get() == user
    assert document.posts.find_by_author_id(1)[0].get() == post

    reopened = BinaryRefDocument.from_bytes(document.to_bytes())
    category_ref = reopened.categorys.get_by_name("Technology")
    assert category_ref is not None
    assert category_ref.get() == category
    assert [item.get() for item in reopened.categorys.search_by_name(" technology ")] == [category]
    assert [item.get() for item in reopened.categorys.search_by_description("reference")] == [category]
    assert [item.get() for item in reopened.categorys.search_by_rank(7)] == [category]
    assert [item.get() for item in reopened.categorys.search_by_kind(TestIndexesCategoryKind.Public)] == [category]
    assert [item.get() for item in reopened.posts.search_by_title("binary")] == [post]
    assert reopened.posts.search_by_title("missing") == []
    assert reopened.tags.get_by_id(200).get() == tag
    assert reopened.post_tags.count() == 1

    invalid_container = BinaryRefContainer(
        categorys=[
            TestIndexesCategory(
                id=11,
                name="Invalid",
                description=None,
                rank=1,
                kind=99,
            )
        ]
    )
    try:
        SchemaBinaryRefContext.save_binary(invalid_container)
    except ValueError:
        pass
    else:
        raise AssertionError("invalid BinaryRef enum discriminant should be rejected on write")

    patched = BinaryRefDocument.from_bytes(patch_category_kind_payload(document, 99))
    try:
        _ = patched.categorys.at(0).kind
    except ValueError:
        pass
    else:
        raise AssertionError("invalid BinaryRef enum discriminant should be rejected on lazy read")


test_indexes_and_foreign_keys()
test_foreign_key_validation_rejects_missing_reference()
test_clear_resets_indexes()
test_enum_csv_json_loaders()
test_container_load_from_sources_csv()
test_container_load_from_sources_json()
test_binary_roundtrip_and_invalid_enum_discriminant()
test_binary_ref_document_roundtrip_and_lazy_indexes()
